### ISSUE TITLE
Deposit contract issue and vanguard validator

### ISSUE THREAD

**Issue by mxmar** [frozeman/vanguard-consensus-engine#41]
# 🐞 Bug Report

### Description

There is a problem with `validator` client in `vanguard`. There is possibility to deploy deposit smart contract, you can interact with contract's methods (`deposit` for example), but `vanguard validator` is unable to fetch info about new deposit and activate validator pubKey to allow vanguard validator client to start confirming/sealing. 

### Has this worked before in a previous version?

The same behavior has been noticed in the previous network state (before creating own consensus from May 2021).

## 🔬 Minimal Reproduction

Connect to `staging` remote `beacon-chain` using `validator` client (from `vanguard` of course). I've built `validator` client by myself, `vanguard` branch is named `feature/process-pending-blocks`. 

## 🔥 Error

<pre><code>
[2021-06-28 16:48:13]  INFO validator: Waiting for deposit to be observed by beacon node pubKey=0x885f65848aaa status=UNKNOWN_STATUS
</code></pre>

## 🌍  Your Environment

**Operating System:**

Local operating system is `macOS Big Sur` (`11.3.1 (20E241)`). Remote operating system is `Ubuntu Server`.

**What version of Prysm are you running? (Which release)**

https://github.com/lukso-network/vanguard-consensus-engine/tree/feature/process-pending-blocks

**Anything else relevant (validator index / public key)?**

Code feedback:
Path: `validator/client/wait_for_activation.go` (`vanguard`).

```
// waitForActivation performs the following:
// 1) While the key manager is empty, poll the key manager until some validator keys exist.
// 2) Open a server side stream for activation events against the given keys.
// 3) In another go routine, the key manager is monitored for updates and emits an update event on
// the accountsChangedChan. When an event signal is received, restart the waitForActivation routine.
// 4) If the stream is reset in error, restart the routine.
// 5) If the stream returns a response indicating one or more validators are active, exit the routine.
```

Validator status feedback:
Path: `beacon-chain/rpc/validator/status.go` (`vanguard`).
```
// ValidatorStatus returns the validator status of the current epoch.
// The status response can be one of the following:
//  DEPOSITED - validator's deposit has been recognized by Ethereum 1, not yet recognized by Ethereum 2.
//  PENDING - validator is in Ethereum 2's activation queue.
//  ACTIVE - validator is active.
//  EXITING - validator has initiated an an exit request, or has dropped below the ejection balance and is being kicked out.
//  EXITED - validator is no longer validating.
//  SLASHING - validator has been kicked out due to meeting a slashing condition.
//  UNKNOWN_STATUS - validator does not have a known status in the network.
```

**Comment by mxmar** [comment]
Go code fragment with validator activation conditions:

```
func validatorStatus(validator *ethpb.Validator, epoch types.Epoch) ethpb.ValidatorStatus {
	farFutureEpoch := params.BeaconConfig().FarFutureEpoch
	if validator == nil {
		return ethpb.ValidatorStatus_UNKNOWN_STATUS
	}
	if epoch < validator.ActivationEligibilityEpoch {
		return ethpb.ValidatorStatus_DEPOSITED
	}
	if epoch < validator.ActivationEpoch {
		return ethpb.ValidatorStatus_PENDING
	}
	if validator.ExitEpoch == farFutureEpoch {
		return ethpb.ValidatorStatus_ACTIVE
	}
	if epoch < validator.ExitEpoch {
		if validator.Slashed {
			return ethpb.ValidatorStatus_SLASHING
		}
		return ethpb.ValidatorStatus_EXITING
	}
	return ethpb.ValidatorStatus_EXITED
}
```

**Comment by mxmar** [comment]
Possible issues - need to debug more:

1. `validatorStatus method searches for the requested validator's state and deposit to retrieve its inclusion estimate. Also returns the validators index.` but it's `Using ^0 as the default value for index, in case the validators index cannot be determined.`

2. `vanguard` is giving invalid `validatorStatus` because of no interaction (or some unexpected error, response from contract) with deposit contract (BE->deposit contract).

3. `validatorStatus` is not working properly -> https://github.com/lukso-network/vanguard-consensus-engine/issues/41#issuecomment-869813721

4. IMHO not possible but maybe some issue between `keyManager` and `validator` client when it checks for `pubKey` activation response.

**Comment by mxmar** [comment]
Next step: find return statement where `vanguard` `beacon-chain` prepares response payload with `UNKNOWN_STATUS` and check why it provides this status code. A this should explain to us everything - condition why it's still `UNKNOWN_STATUS`.

**Comment by atif-konasl** [comment]
@mxmar Currently our codebase is far away from prysm's recent version(`v1.3.11`). Prysm have implemented lot of features as well as fixing bugs. There could be some issue in our code base. I think it would be great if you test this deposit contract feature using prysm's current release because when we will finish our all features, we will merge with updated prysm's codebase.

**Comment by mxmar** [comment]
I was debugging with `endtoend/endtoend_test.go` in our `vanguard` implementation and with the latest `prysm` implementation. In our case it's failing but in case of `prysm` it is working. One of the differences was that e2e tests are using `go-ethereum` (geth) as mock for ETH1 chain.

**Comment by mxmar** [comment]
The problem is that we are new network (new `prysm` and new `geth` networks).  We are the only guys who are facing this issue.

**Comment by mxmar** [comment]
`Unknown status means the validator has not been put into the state yet.` it's about -> `ethpb.ValidatorStatus_UNKNOWN_STATUS`

**Comment by mxmar** [comment]
`ProcessRegistryUpdates` from `beacon-chain/core/epoch/epoch_processing.go` -> rotates validators in and out of active pool, the amount to rotate is determined churn limit.

**Comment by mxmar** [comment]
`beacon-chain/powchain/log_processing.go` and:

```
if err := s.processDeposit(ctx, eth1Data, deposit); err != nil {
			log.Errorf("Invalid deposit processed: %v", err)
			validData = false
		}
```

Issue should lies there.

**Comment by mxmar** [comment]
This is log from `remix` after `deposit` transact happened:

```
[
	{
		"from": "0x000000000000000000000000000000000000cafE",
		"topic": "0x649bbc62d0e31342afea4e5cd82d4049e7e1ee912fc0889aa790803be39038c5",
		"event": "DepositEvent",
		"args": {
			"0": "0x947b75e821927ff3c7ca38de8c9f0e855ccdd42ed705c900d153c9dc730e79d6f9afc2551db29b7cb2f45061cd60c6db",
			"1": "0x010000000000000000000000cedbeb55f9f439ad482a90076b59c2823696f925",
			"2": "0x0040597307000000",
			"3": "0xad8c60ec0c9cc105e6f59a1e38d7e0254bb6f3eb00474f9c34fed0da45b78eb2494f6411dd196a9130e2e7b6ea114a561929b53d7f43094e48b5889db8d54bd6696f08ef4d0a925671e14808dde51d3ec3f9373460fa8ea764bd83bd6bf9b610",
			"4": "0x0100000000000000",
			"pubkey": "0x947b75e821927ff3c7ca38de8c9f0e855ccdd42ed705c900d153c9dc730e79d6f9afc2551db29b7cb2f45061cd60c6db",
			"withdrawal_credentials": "0x010000000000000000000000cedbeb55f9f439ad482a90076b59c2823696f925",
			"amount": "0x0040597307000000",
			"signature": "0xad8c60ec0c9cc105e6f59a1e38d7e0254bb6f3eb00474f9c34fed0da45b78eb2494f6411dd196a9130e2e7b6ea114a561929b53d7f43094e48b5889db8d54bd6696f08ef4d0a925671e14808dde51d3ec3f9373460fa8ea764bd83bd6bf9b610",
			"index": "0x0100000000000000"
		}
	}
]
```

**Comment by mxmar** [comment]
Diagram for `deposit` activation flow, how `beacon-chain` interacts with `depositEvent` -> https://miro.com/app/board/o9J_l8DoFqk=/

**Comment by mxmar** [comment]
Deposit contract issue has been solved by `non-interop` network. We can't use `interop` flags in `validator` and `beacon-chain` (`vanguard`) to have proper flow of `deposit`. We had to use `non-interop` network.

https://github.com/lukso-network/vanguard-consensus-engine/tree/chore/minimal-consensus-info (contains consensus info fixes)
https://github.com/lukso-network/vanguard-consensus-engine/tree/release/non-interop (doesn't contain consensus info fixes)

PR: https://github.com/lukso-network/vanguard-consensus-engine/pull/45

Spike branch used to debug `deposit contract` issue was https://github.com/lukso-network/vanguard-consensus-engine/tree/spike/non-interop-network

**Summary - what is done or needs to be done:**
- [x] with `non-interop` network there are no issues with `genesis.ssz` with `deposit-data.json` keys inserted. `validators` activated by `genesis.ssz` list;
- [x] with `non-interop` network there were no issues with `deposit` flow -> when you send 32ETH deposit transaction to the `deposit contract` -> it was possible to activate validator but it takes some time (e.g. 5-30mins, depends on network state and eth1 logs size).
- [x] when we achieve proper chain progression with `ephemeral test network` then we need to test `deposit` flow again -> flow with 32ETH transaction. `genesis.ssz` flow is working at all.
